sungrow/integrationpluginsungrow.cpp

```diff
@@ -0,0 +1,375 @@
+                    child->setStateValue(sungrowMeterCurrentPhaseCStateTypeId, 0);
+                }
+
+                child = getBatteryThing(thing);
+                if (child) {
+                    child->setStateValue(sungrowBatteryCurrentPowerStateTypeId, 0);
+                    child->setStateValue(sungrowBatteryBatteryLevelStateTypeId, 0);
+                    child->setStateValue(sungrowBatteryBatteryCriticalStateTypeId, true);
+                    child->setStateValue(sungrowBatteryChargingStateStateTypeId, "idle");
+                }
+            }
+        });
+
+        connect(sungrowConnection, &SungrowModbusTcpConnection::initializationFinished, thing, [=](bool success){
+            thing->setStateValue("connected", success);
+
+            foreach (Thing *childThing, myThings().filterByParentId(thing->id())) {
+                childThing->setStateValue("connected", success);
+            }
+
+            if (!success) {
+                // Try once to reconnect the device
+                sungrowConnection->reconnectDevice();
+            } else {
+                qCInfo(dcSungrow()) << "Connection initialized successfully for" << thing;
+                sungrowConnection->update();
+            }
+        });
+
+        connect(sungrowConnection, &SungrowModbusTcpConnection::updateFinished, thing, [=](){
+            qCDebug(dcSungrow()) << "Updated" << sungrowConnection;
+
+            if (myThings().filterByParentId(thing->id()).filterByThingClassId(sungrowMeterThingClassId).isEmpty()) {
+                qCDebug(dcSungrow()) << "There is no meter set up for this inverter. Creating a meter for" << thing << sungrowConnection->modbusTcpMaster();
+                ThingClass meterThingClass = thingClass(sungrowMeterThingClassId);
+                ThingDescriptor descriptor(sungrowMeterThingClassId, meterThingClass.displayName() + " " + sungrowConnection->serialNumber(), QString(), thing->id());
+                emit autoThingsAppeared(ThingDescriptors() << descriptor);
+            }
+
+            // Check if a battery is connected to this Sungrow inverter
+            if (sungrowConnection->batteryType() != SungrowModbusTcpConnection::BatteryTypeNoBattery &&
+                    myThings().filterByParentId(thing->id()).filterByThingClassId(sungrowBatteryThingClassId).isEmpty()) {
+                qCDebug(dcSungrow()) << "There is a battery connected but not set up yet. Creating a battery.";
+                ThingClass batteryThingClass = thingClass(sungrowBatteryThingClassId);
+                ThingDescriptor descriptor(sungrowBatteryThingClassId, batteryThingClass.displayName() + " " + sungrowConnection->serialNumber(), QString(), thing->id());
+                emit autoThingsAppeared(ThingDescriptors() << descriptor);
+            }
+
+            // Update inverter states
+            thing->setStateValue(sungrowInverterTcpCurrentPowerStateTypeId, sungrowConnection->totalPVPower());
```

Comment:
After testing the code, the sign of the inverter power is still wrong. `-sungrowConnection->totalPVPower()`

Reply:
Added ( .... * -1) to make it more obvious to the casual reader.